plugins/formal-verification/Uniqueness.md

```diff
@@ -0,0 +1,187 @@
+meaningful to speak of aliasing occurring within some context; if two instance variables refer to a single object, but
+one of them belongs to an object that cannot be reached from anywhere in the system, then the
+aliasing is irrelevant.
+
+Within any method, objects may be accessed through paths rooted at any of:
+
+- Self
+- An anonymous locally constructed object
+- A method argument
+- A result returned by another method
+- A (global) variable accessible from the method scope
+- A local method variable bounded to any of the above
+
+An object is **aliased** with respect to some context if two or more such paths to it exist.
+
+# Aliasing in Kotlin
+
+## Smart casts
+
+The aliasing issue is evident in Kotlin, as the language does not impose any restrictions on it. In the following
+example, the compiler is unable to execute a seemingly obvious smart cast from the user's perspective. However, the
+compiler's analysis is accurate, as `x` may be aliased, and another thread could potentially modify the `x.n` field.
+
+```kt
+class X(var n: Int?)
+
+fun useX(x: X): Int =
+    if (x.n != null) {
+        x.n
+//      ^^^ 
+// Smart cast to 'Int' is impossible, because 'x.n' is a mutable property that could have been changed by this time
+    } else {
+        0
+    }
+```
+
+## Proving functional behaviour
+
+The previous example illustrates how even basic functional properties can't be validated when aliasing and mutability
+are involved. Aliasing in Kotlin also poses a challenge for the static analyzer used by IntelliJ IDEA. In the following
+example, aliasing results in an incorrect constant condition as provided by the IDE.
+
+```kt
+class A(var a: Boolean = false)
+
+fun f(a1: A, a2: A) {
+    a1.a = true
+    a2.a = false
+    if (!a1.a) {
+//      ^^^^^
+// Condition '!a1.a' is always false 
+        println("ALIASED!")
+    }
+}
+
+fun main() {
+    val a1 = A()
+    f(a1, a1) // prints "ALIASED!"
+}
+```
+
+# Our Uniqueness System
+
+We propose an annotation system able to provide some guarantees on the uniqueness of the variables.
+
+## Uniqueness invariant
+
+A unique variable is one whose value is null or else refers to an unshared object, one referred to by no other
+variables. This situation is called the _uniqueness invariant_. Parameters, receivers and return values may be declared
+unique. A use of a unique variable is a destructive read: the variable is atomically set to null at the same time that
+the value is read. The destruction preserves the _uniqueness invariant_.
+
+## Alias Burying and relaxation of uniqueness invariant
+
+Ensuring the _uniqueness invariant_ requires unique variables to be set null every time they are
+read. [Alias burying][2] keeps the uniqueness invariant but only require it to be true when it is needed. So the
+uniqueness invariant is thus not actually true at every point in the program, but the points when it is false are
+‘uninteresting’.
```

Comment:
Can the cases that it is interesting be summed up in a sentence?  It's a bit vague otherwise.